Samples/Hast.Samples.Consumer/Gui.cs

```diff
@@ -0,0 +1,356 @@
+using Hast.Layer;
+using Hast.Samples.Consumer.Models;
+using Microsoft.Extensions.Logging;
+using Newtonsoft.Json;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Terminal.Gui;
+
+namespace Hast.Samples.Consumer
+{
+    public class Gui
+    {
+        private readonly Dictionary<string, ConsumerConfiguration> _savedConfigurations;

```

Comment:
Nem veszi be a device configot

https://user-images.githubusercontent.com/1976647/139743603-02a433e3-e04d-47a7-a55c-c0764af0343e.mp4

:

Reply:
ír a log valami exceptiont?

Reply:
Erre nem. A később emiatti exceptiont loggolja:

```
2021-11-01 22:12:11.0430|Hast.Layer.Hastlayer|ERROR|An error happened during generating the Hastlayer hardware representation for the following assemblies: E:\Projects\Lombiq\Hastlayer\Hastlayer-SDK\Samples\Hast.Samples.Consumer\bin\Debug\netcoreapp3.1\Hast.Samples.SampleAssembly.dll, E:\Projects\Lombiq\Hastlayer\Hastlayer-SDK\Samples\Hast.Samples.Consumer\bin\Debug\netcoreapp3.1\Hast.Algorithms.dll, E:\Projects\Lombiq\Hastlayer\Hastlayer-SDK\Samples\Hast.Samples.Consumer\bin\Debug\netcoreapp3.1\Hast.Samples.FSharpSampleAssembly.dll, E:\Projects\Lombiq\Hastlayer\Hastlayer-SDK\Samples\Hast.Samples.Consumer\bin\Debug\netcoreapp3.1\Lombiq.Arithmetics.dll, E:\Projects\Lombiq\Hastlayer\Hastlayer-SDK\Samples\Hast.Samples.Consumer\bin\Debug\netcoreapp3.1\System.Collections.Immutable.dll, E:\Projects\Lombiq\Hastlayer\Hastlayer-SDK\Samples\Hast.Samples.Consumer\bin\Debug\netcoreapp3.1\FSharp.Core.dll System.InvalidOperationException: XILINX_VITIS variable is not set. This is required to run Vivado. For further instructions see https://www.xilinx.com/html_docs/xilinx2020_1/vitis_doc/settingupvitisenvironment.html
   at Hast.Vitis.Abstractions.Services.VitisHardwareImplementationComposerBuildProvider.BuildAsync(IHardwareImplementationCompositionContext context, IHardwareImplementation implementation) in E:\Projects\Lombiq\Hastlayer\Hastlayer-SDK\Hast.Abstractions\Hast.Vitis.Abstractions\Services\VitisHardwareImplementationComposerBuildProvider.cs:line 104
   at Hast.Synthesis.Abstractions.VhdlHardwareImplementationComposer.ComposeAsync(IHardwareImplementationCompositionContext context) in E:\Projects\Lombiq\Hastlayer\Hastlayer-SDK\Hast.Abstractions\Hast.Synthesis.Abstractions\VhdlHardwareImplementationComposer.cs:line 54
   at Hast.Layer.Hastlayer.GenerateHardware(IEnumerable`1 assemblyPaths, IHardwareGenerationConfiguration configuration) in E:\Projects\Lombiq\Hastlayer\Hastlayer-SDK\Hast.Layer\Hastlayer.cs:line 251    at Hast.Vitis.Abstractions.Services.VitisHardwareImplementationComposerBuildProvider.BuildAsync(IHardwareImplementationCompositionContext context, IHardwareImplementation implementation) in E:\Projects\Lombiq\Hastlayer\Hastlayer-SDK\Hast.Abstractions\Hast.Vitis.Abstractions\Services\VitisHardwareImplementationComposerBuildProvider.cs:line 104
   at Hast.Synthesis.Abstractions.VhdlHardwareImplementationComposer.ComposeAsync(IHardwareImplementationCompositionContext context) in E:\Projects\Lombiq\Hastlayer\Hastlayer-SDK\Hast.Abstractions\Hast.Synthesis.Abstractions\VhdlHardwareImplementationComposer.cs:line 54
   at Hast.Layer.Hastlayer.GenerateHardware(IEnumerable`1 assemblyPaths, IHardwareGenerationConfiguration configuration) in E:\Projects\Lombiq\Hastlayer\Hastlayer-SDK\Hast.Layer\Hastlayer.cs:line 251
```

Reply:
Tudod csatolni a ConsumerConfiguration.json-t?
Már csak ez, meg az egérrel kapcsolatos UX probléma van, azt valamikor átbeszélhetnénk

Reply:
(.json fájlt nem lehet feltölteni)

[ConsumerConfiguration.txt](https://github.com/Lombiq/Hastlayer-SDK/files/7491459/ConsumerConfiguration.txt)

Reply:
```json
{
  "test": {
    "DeviceName": null,
    "Endpoint": null,
    "AppName": null,
    "AppSecret": null,
    "SampleToRun": 0,
    "BuildLabel": null,
    "DontRun": false,
    "VerifyResults": false,
    "HardwareFrameworkPath": "HardwareFramework"
  }
}
```

Hát ez tök üres, szóval a betöltéssel nincs gond. Talán a mentés ment félre. Mit akartál bele menteni?

Reply:
Kiválasztottam a Nexyst, mint device-t, majd rányomtam a Startra.

Reply:
Meg is kell erősíteni enter lenyomásával, mert különben a billentyűzetes kiválasztásnál minden egyes lépésnél átírná ami fura lenne. Hozzáadtam egy szöveget az aljára.
![image](https://user-images.githubusercontent.com/4007293/140654940-1df1cd4d-7604-4dea-bab3-449e10f4a5c2.png)

Reply:
Ez így eléggé counter-intuitive. Miért ne menthetnénk rögtön a kiválasztáskor?

Reply:
Azért, hogy egységesen viselkedjen a list view és a text input.

Reply:
Ugye az a cél, hogy ne lehessen a state-et véletlenül szerkeszteni, illetve szövegnél félkész állapotba hagyni.

Most hozzáadtam egy Confirm gombot, szerintem így már egyértelmű, hogy szándékos megerősítés kell:
![image](https://user-images.githubusercontent.com/4007293/140668317-588488a3-53fc-4050-b865-4bec9491bccb.png)

Reply:
A text inputnak is onfocusoutra mentenie kellene. Ha ez egy WPF GUI lenne, akkor se várnánk el minden lapon egy Save gomb megnyomását. Lehet ez valami átmeneti módon tárolva is (mert az állandóra végül is ott van az explicit save/load), a lényeg, hogy ha valamit összekattintgatsz és rányomsz a Start-ra, akkor működjön, ne legyen külön, pláne ne minden opcionál egy külön kötelező mentés, mert az szerintem elég macerás.

Reply:
Hja, de ez sajnos nem WPF. Ha reaktív lenne, nem lenne fele ilyen macerás dolgozni vele... 😢

Reply:
Nagy nehezen sikerült, de eléggé agyrohasztó volt a logika.

Reply:
Nice!

Reply:
Ez most úgy van, hogy ha egy opciót kiválasztasz és Start, akkor OK. Ha opciót kiválasztasz, majd átmész egy másik lapra, ott pedig egy másik opciót, akkor az elsőt elfelejti.

Reply:
Hmm, a `ListView` nem triggereli a `Leave` eventet (ez az on focus out), de a `TextField` igen. Szerintem ez library bug lesz. [Írtam itt](https://github.com/migueldeicaza/gui.cs/issues/1507). Szerintem össze tudnék hekkelni valami workaroundot, de inkább megvárom, hogy mit írnak.

Reply:
Ahha. Érdemes lett volna akkor videót csatolnom, de hát annyira nyilvánvaló volt a helyzet, hogy minek, ugye.

Reply:
Végre bepusholták. De nincs nightly/pre-release NuGetük és ki tudja mikor lenne a következő release, ezért inkább beraktam submodule-ként. Csináljak future issuet rá, hogy ha kijön az 1.3.2 akkor rakjuk vissza NuGetre?

Reply:
Szokott viszonylag gyakran lenni. Vájunk egy hetet, ha nincs release, akkor submodule és issue, ja.

Reply:
Ok, várjunk. De én az RC-ket leszámítva nem látom, hogy olyan nagyon gyakran lenne release:
![image](https://user-images.githubusercontent.com/4007293/141661678-2ed89292-feca-479d-b539-0e302277ce6e.png)

Reply:
No release :(

Reply:
Akkor submodule és issue.